gh-119698: fix `symtable.Class.get_methods` and document its behaviour correctly

picnixz opened the pull request:
This is an alternative to #120148.

<!-- gh-issue-number: gh-119698 -->
* Issue: gh-119698
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--120151.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

picnixz committed:
add test

picnixz committed:
fix `symtable.Class.get_methods`

picnixz committed:
add doc

picnixz committed:
blurb

picnixz committed:
fixup

picnixz committed:
fixup whitespaces

picnixz committed:
fixup

picnixz committed:
fixup

carljm reviewed (changes requested):
I think we should also test here that we do identify a PEP 695 generic method (e.g. `def f[T](self): ...`) as a method.

(I don't have strong opinions about whether we fix or document `get_methods`, but seems like we may as well fix if this much of the work is already done to fix.)

picnixz committed:
improve test coverage

picnixz committed:
Merge branch 'main' into fix-symtable-get-methods

JelleZijlstra commented on Lib/test/test_symtable.py:
```diff
@@ -51,6 +51,67 @@ def generic_spam[T](a):
+    global glob_assigned_meth_ignore
+    def glob_assigned_meth_ignore(): pass
+    global glob_assigned_meth_pep_695_ignore
+    def glob_assigned_meth_pep_695_ignore[T](): pass
```
Also add an `async def` method.

picnixz committed:
add async def cases

picnixz committed:
update comments

JelleZijlstra commented on Lib/symtable.py:
```diff
@@ -221,8 +221,27 @@ def get_methods(self):
+                        # have more in the future). In particular, we need
+                        # to find the corresponding inner function, class or
+                        # type alias.
+                        st = next((c for c in st.children if c.name == st.name), None)
```
I'm not sure this will keep working once PEP 649 is implemented since I think I gave the annotation scope the same name. (Don't have time to verify right now.)

picnixz replied:
I'll do something like that then (I'm no more on Linux, so I'll make the changes tomorrow):

```py
if st.type == TYPE_TYPE_PARAM:  # or TYPE_TYPE_PARAMS in the future
	for child in st.children:
	    if child.name == st.name and child.type == TYPE_TYPE_FUNCTION:
			d[st.name] = 1
			break
elif st.type == TYPE_TYPE_FUNCTION:
	d[st.name] = 1
```

picnixz committed:
prepare for handling PEP 649

picnixz committed:
unapply restyle

carljm reviewed (approved):
Looks good to me!

picnixz committed:
improve test

- extract the 'ComplexClass' test case from simple test code
- rename some variables in the test code

picnixz committed:
normalize declaration order

picnixz committed:
improve example

picnixz committed:
Merge branch 'main' into fix-symtable-get-methods

picnixz committed:
Merge branch 'main' into fix-symtable-get-methods

picnixz committed:
Merge branch 'main' into fix-symtable-get-methods

# Conflicts:
#	Lib/symtable.py

JelleZijlstra commented on Doc/library/symtable.rst:
```diff
@@ -127,8 +127,39 @@ Examining Symbol Tables
+      body via :keyword:`def` or :keyword:`async def`.
+
+      Functions defined in a deeper scope (e.g., in an inner class) are not
+      picked by :meth:`get_methods`.
```
```suggestion
      picked up by :meth:`get_methods`.
```

picnixz committed:
Update Doc/library/symtable.rst

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

JelleZijlstra commented on Lib/symtable.py:
```diff
@@ -221,10 +221,25 @@ def get_methods(self):
         """
         if self.__methods is None:
             d = {}
```
Can we use a set here instead? I feel this code may predate the addition of sets to the language.

picnixz replied:
With a set, I would lose the order of discovery, so I think a dict is still good.

picnixz committed:
Merge branch 'main' into fix-symtable-get-methods

picnixz committed:
Merge branch 'main' into fix-symtable-get-methods

picnixz committed:
fix typo...
